presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PreferredProperties.java

```diff
@@ -96,6 +100,70 @@ public static PreferredProperties local(List<? extends LocalProperty<Symbol>> lo
         return localProperties;
     }
 
+    public static PreferredProperties derivePreferences(PreferredProperties parentProperties,
+            Set<Symbol> partitioningColumns,
+            List<LocalProperty<Symbol>> localProperties)
+    {
+        // Build identity translations
+        ImmutableMap.Builder<Symbol, Symbol> translations = ImmutableMap.builder();
+        partitioningColumns.forEach(sym -> translations.put(sym, sym));
+        return derivePreferences(parentProperties, translations.build(), Optional.of(partitioningColumns), Optional.empty(), localProperties);
+    }
+
+    public static PreferredProperties derivePreferences(PreferredProperties parentProperties,
+            Map<Symbol, Symbol> translations,
+            Optional<Set<Symbol>> partitioningColumns,
+            Optional<List<Symbol>> hashingColumns,
+            List<LocalProperty<Symbol>> localProperties)
+    {
+        List<LocalProperty<Symbol>> parentLocalProperties = LocalProperties.translate(parentProperties.getLocalProperties(), column -> Optional.ofNullable(translations.get(column)));
+
+        // if the child plan can first meet our requirements and then by our parent's,
+        // it can satisfy both in one shot
+        List<LocalProperty<Symbol>> local = ImmutableList.<LocalProperty<Symbol>>builder()
+                .addAll(localProperties)
+                .addAll(parentLocalProperties)
+                .build();
+
+        // Check we need to be hash partitioned
+        if (hashingColumns.isPresent() && partitioningColumns.isPresent() && partitioningColumns.get().equals(ImmutableSet.copyOf(hashingColumns.get()))) {
+            return PreferredProperties.hashPartitionedWithLocal(hashingColumns.get(), local);
```

Comment:
Does `hashingColumns` represent the output or input columns of the operator for which we're trying to derive the properties? If it's the output columns, then this needs to translate them.

In any case, I think we need to clarify a bit what each of the arguments to this method represents (inputs vs outputs, etc)